ports/esp32/esp_espnow.c

```diff
@@ -0,0 +1,315 @@
+    esp_now_peer_info_t peer;
+    ESPNOW_EXCEPTIONS(esp_now_get_peer(addr_buf, &peer));
+
+    // set peer lmk
+    bool encrypt = (key != mp_const_none);
+    bool re_add = (peer.encrypt != encrypt);
+    if (encrypt) _get_bytes(key, ESP_NOW_KEY_LEN, peer.lmk);
+    if (re_add) {
+        // workaround for calling esp_now_mod_peer() to 
+        // change encryption status will crash the system
+        peer.encrypt = encrypt;
+        ESPNOW_EXCEPTIONS(esp_now_del_peer(addr_buf));
+        ESPNOW_EXCEPTIONS(esp_now_add_peer(&peer));
+    } else {
+        ESPNOW_EXCEPTIONS(esp_now_mod_peer(&peer));
+    }
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(espnow_lmk_obj, espnow_lmk);
+
+// add_peer(peer_mac, [local_key])
+STATIC mp_obj_t espnow_add_peer(size_t n_args, const mp_obj_t *args) {
+    esp_now_peer_info_t peer = {0};
+    _get_bytes(args[0], ESP_NOW_ETH_ALEN, peer.peer_addr);
+    if (n_args > 1) {
+        _get_bytes(args[1], ESP_NOW_KEY_LEN, peer.lmk);
+        peer.encrypt = 1;
+    }
+
+    ESPNOW_EXCEPTIONS(esp_now_add_peer(&peer));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(espnow_add_peer_obj, 1, 2, espnow_add_peer);
+
+// del_peer(peer_mac)
+STATIC mp_obj_t espnow_del_peer(mp_obj_t addr) {
+    mp_uint_t addr_len;
+    const uint8_t *addr_buf = (const uint8_t *)mp_obj_str_get_data(addr, &addr_len);
+    if (addr_len != ESP_NOW_ETH_ALEN) mp_raise_ValueError("addr invalid");
+    ESPNOW_EXCEPTIONS(esp_now_del_peer(addr_buf));
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(espnow_del_peer_obj, espnow_del_peer);
+
+// this workaround enables ESP32 to send from whatever IF that is
+// active
+// if_id == wifi_mode - 1
+#define IS_IF_AVAILABLE(mode, if_id) ({ (mode & (if_id+1)) != 0; })
+#define AVAILABLE_IF(mode) ({                                       \
+    int if_id = -1;                                                 \
+    for (int i=WIFI_MODE_STA; i<=WIFI_MODE_AP; i++) {               \
+        if (mode & i) {                                             \
+            if_id = i-1;                                            \
+        }                                                           \
+    }                                                               \
+    if_id;                                                          \
+})                                                                  \
+
+STATIC mp_obj_t espnow_send(mp_obj_t addr, mp_obj_t msg) {
+    if (!wifi_started) goto espnow_wifi_err;
+
+    mp_uint_t addr_len;
+    const uint8_t *addr_buf;
+    mp_uint_t msg_len;
+    const uint8_t *msg_buf = (const uint8_t *)mp_obj_str_get_data(msg, &msg_len);
+    if (msg_len > ESP_NOW_MAX_DATA_LEN) mp_raise_ValueError("msg too long");
+
+    wifi_mode_t mode;
+    ESPNOW_EXCEPTIONS(esp_wifi_get_mode(&mode));
+    bool first = true;
+    int new_if = -1;
+    if (addr == mp_const_none) {
+        // send to all
```

Comment:
esp32: hmm, this does not look like a broadcast (in the sense of "everybody, not just registered peers"), right?

see also the esp8266 implementation, which seems different.

is this a esp32 limitation?

Reply:
found the answer:

first one needs to explicitly add the broadcast "peer" like a unicast peer (so it will be in the list of peers):

https://docs.espressif.com/projects/esp-idf/en/latest/api-reference/network/esp_now.html#add-paired-device

Reply:
there is still the question somehow whether this is different between esp32 and esp8266 as the docs seem to describe the api as a unified thing for esp32 and esp8266, but this code deals differently with the "None/broadcast" case depending on the device.

Reply:
> esp32: hmm, this does not look like a broadcast (in the sense of "everybody, not just registered peers"), right?
> 
> see also the esp8266 implementation, which seems different.
> 
> is this a esp32 limitation?

What you are looking at is actually a workaround for a bug(?) in the api.

esp32: 
When registering a peer, that peer will automatically bind to an available IF.
If, for some reason, that IF is down at the time when sending messages, then message won't be sent.

These extra code just making sure that message will be sent through as long as one IF is available.

Reply:
> there is still the question somehow whether this is different between esp32 and esp8266 as the docs seem to describe the api as a unified thing for esp32 and esp8266, but this code deals differently with the "None/broadcast" case depending on the device.

Being a proprietary protocol, they differ drastically. I typically call the esp32 version of espnow a V2.